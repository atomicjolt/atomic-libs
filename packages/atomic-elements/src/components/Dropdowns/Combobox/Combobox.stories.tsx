import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { TextInputControls } from "@sb/helpers";
import { ComboBox } from ".";
import { Item } from "../../Collection";
import { getCssProps } from "@sb/cssprops";

const meta: Meta<typeof ComboBox> = {
  title: "Dropdowns/Selection/Combobox",
  component: ComboBox,
  parameters: {
    layout: "centered",
    cssprops: getCssProps("Dropdown", "Input", "Listbox"),
  },
  argTypes: {
    ...TextInputControls,
    variant: {
      table: { category: "Presentation" },
    },
    children: { control: false },
    allowsCustomValue: {
      control: "boolean",
      description:
        "Whether the ComboBox allows the user to enter a custom value. If true, the ComboBox will allow the user to enter a value that is not in the list of options. If false, the ComboBox will only allow the user to select from the list of options.",
    },
    defaultInputValue: {
      control: "text",
      description:
        "The default value of the input field for an uncontrolled comobox",
    },
    inputValue: {
      control: "text",
      description: "The value of the input field for a controlled combobox.",
    },
    defaultSelectedKey: {
      control: "select",
      options: ["1", "2"],
      description: "Key of the selected item for an uncontrolled component.",
    },
    selectedKey: {
      control: "select",
      options: ["1", "2"],
      description: "Key of the selected item for a controlled component.",
    },
    disabledKeys: {
      control: "multi-select",
      options: ["1", "2", "3"],
      description: "Array of the disabled keys.",
    },
    menuTrigger: {
      control: "select",
      options: ["manual", "input", "focus"],
      description:
        "Controls when the dropdown menu is shown. `manual` will only show the dropdown when the user clicks the button. `input` will show the dropdown when the input is focused & the user enters input. `focus` will show the dropdown when the input is focused",
    },
  },
};

export default meta;

type Story = StoryObj<typeof ComboBox>;

export const Primary: Story = {
  args: {
    label: "Combobox label",
    menuTrigger: "focus",
    children: [
      <Item key="1" id="1">
        Option 1
      </Item>,
      <Item key="2" id="2">
        Option 2
      </Item>,
      <Item key="3" id="3">
        Option 3
      </Item>,
    ],
  },
};

export const Floating: Story = {
  args: {
    ...Primary.args,
    variant: "floating",
  },
};

export const PrefixIcon: Story = {
  args: {
    ...Primary.args,
    prefixIcon: "search",
    icon: "arrow_drop_down",
    size: "large",
  },
};

export const ScollingDropdown: Story = {
  args: {
    ...Primary.args,
    children: [
      <Item key="1" id="1">
        Option 1
      </Item>,
      <Item key="2" id="2">
        Option 2
      </Item>,
      <Item key="3" id="3">
        Option 3
      </Item>,
      <Item key="4" id="4">
        Option 4
      </Item>,
      <Item key="5" id="5">
        Option 5
      </Item>,
      <Item key="6" id="6">
        Option 6
      </Item>,
      <Item key="7" id="7">
        Option 7
      </Item>,
      <Item key="8" id="8">
        Option 8
      </Item>,
      <Item key="9" id="9">
        Option 9
      </Item>,
      <Item key="10" id="10">
        Option 10
      </Item>,
      <Item key="11" id="11">
        Option 11
      </Item>,
      <Item key="12" id="12">
        Option 12
      </Item>,
      <Item key="13" id="13">
        Option 13
      </Item>,
      <Item key="14" id="14">
        Option 14
      </Item>,
      <Item key="15" id="15">
        Option 15
      </Item>,
      <Item key="16" id="16">
        Option 16
      </Item>,
      <Item key="17" id="17">
        Option 17
      </Item>,
      <Item key="18" id="18">
        Option 18
      </Item>,
      <Item key="19" id="19">
        Option 19
      </Item>,
      <Item key="20" id="20">
        Option 20
      </Item>,
    ],
  },
};

export const DynamicCollection: Story = {
  render: () => {
    let options = [
      { id: 1, name: "Aerospace" },
      { id: 2, name: "Mechanical" },
      { id: 3, name: "Civil" },
    ];

    return (
      <ComboBox label="Pick a engineering major" items={options}>
        {(item) => <Item>{item.name}</Item>}
      </ComboBox>
    );
  },
  args: {
    ...Primary.args,
  },
};
