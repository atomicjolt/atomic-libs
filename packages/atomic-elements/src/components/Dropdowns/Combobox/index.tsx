export { ComboBox } from "./ComboBox.component";
export type { ComboBoxProps } from "./ComboBox.component";
