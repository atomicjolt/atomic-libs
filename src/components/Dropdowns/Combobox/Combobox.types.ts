import { SharedInputProps } from "../../../types";

export interface ComboboxProps extends SharedInputProps {
  value?: string;
  onChange?: (value: string) => void;

  /** Array of possible values to suggest to the user */
  options: string[];

  /** Optional function to override the filtering behavior. By default,
   * the suggestions are compared to the current input value by string value.
   * It checks if the suggestions starts with the current value OR if the
   * levenshtein distance between them is 2 or less.
   */
  filterSuggestions?: (value: string, options: string[]) => string[];
}
