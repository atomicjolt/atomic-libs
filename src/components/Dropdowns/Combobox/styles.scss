@use "../../defines" as *;
@use "../common.scss";
