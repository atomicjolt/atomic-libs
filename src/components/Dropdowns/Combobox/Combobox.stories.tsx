import { ComponentMeta, ComponentStory } from "@storybook/react";
import Combobox, { Props } from ".";

export default {
title: "Dropdowns/Combobox",
component: Combobox
} as ComponentMeta<typeof Combobox>

const Template: ComponentStory<typeof Combobox> = (args: Props) => (
  <Combobox {...args}>
    <span>Option 1</span>
  </Combobox>
);

export const Default = Template.bind({});
Default.args = {
  label: "Combobox label",
}

// Additional stories go here