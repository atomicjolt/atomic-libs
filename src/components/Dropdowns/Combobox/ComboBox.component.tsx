import { AriaComboBoxProps } from "@react-aria/combobox";

import {
  AriaProps,
  CanHaveIcon,
  DropdownProps,
  FieldInputProps,
  HasVariant,
  MaterialIcons,
} from "../../../types";
import { IconButton } from "../../Buttons/IconButton";
import { Popover } from "../../Overlays/Popover";
import { ListBox } from "../ListBox";
import { ComboInput, Input } from "../../Fields";
import { MaterialIcon } from "@components/Icons/MaterialIcon";
import { FloatingFieldInputWrapper } from "@components/Internal/FloatingFieldInputWrapper";
import { StyledComboBoxField } from "./Combobox.styles";

export interface ComboBoxProps<T>
  extends AriaProps<AriaComboBoxProps<T>>,
    FieldInputProps,
    CanHaveIcon,
    HasVariant<"default" | "floating">,
    DropdownProps {
  /** Optional icon to render before the input */
  prefixIcon?: MaterialIcons;
}

/** Combox combinds a text input field with a dropdown list of options for the user to select from */
import React, { forwardRef } from "react";

export const ComboBox = forwardRef(function ComboBox<T extends object>(
  props: ComboBoxProps<T>,
  ref: React.Ref<HTMLInputElement>
) {
  const {
    icon = "search",
    prefixIcon,
    iconVariant,
    label,
    message,
    error,
    variant = "default",
    maxHeight = 300,
    dropdownPlacement = "bottom start",
  } = props;

  return (
    <StyledComboBoxField {...props} ref={ref}>
      <FloatingFieldInputWrapper
        label={label}
        message={message}
        error={error}
        floating={variant === "floating"}
      >
        <ComboInput padding={["left", "right"]}>
          {prefixIcon && (
            <MaterialIcon icon={prefixIcon} variant={iconVariant} />
          )}
          <Input size="full" />
          <IconButton
            icon={icon}
            variant="content"
            iconVariant={iconVariant}
            size="auto"
          />
        </ComboInput>
      </FloatingFieldInputWrapper>

      <Popover maxHeight={maxHeight} placement={dropdownPlacement}>
        <ListBox items={props.items}>{props.children}</ListBox>
      </Popover>
    </StyledComboBoxField>
  );
}) as <T extends object>(
  props: ComboBoxProps<T> & { ref?: React.Ref<HTMLInputElement> }
) => JSX.Element;
