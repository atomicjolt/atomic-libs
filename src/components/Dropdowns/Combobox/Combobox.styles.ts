import styled from "styled-components";
import mixins from "../../../styles/mixins";
import { DropdownWrapper } from "../Dropdowns.styles";

export const ComboboxWrapper = styled(DropdownWrapper)`
  ${mixins.SizingX}
  ${mixins.InputWrapper}

  &.read-only {
    --input-bg-clr: var(--neutral100);
  }
`;
