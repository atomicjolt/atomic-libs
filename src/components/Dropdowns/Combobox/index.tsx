import React, { useState, useEffect } from 'react'
import '../../general.scss'
import './styles.scss'

export interface Props {
  children: React.ReactNode;
  /** Must include a label. Labels are always Sentence case. */
  label: string;
  /** Error text should be descriptive and explicit in meaning. */
  error?: string;
  /** For additional information (ex. date format mm/dd/yy) */
  message?: string;
  /** Only use in very specific circumstances. This hides the label from view, but still allows screen readers to read the label. (A filter dropdown with a clear meaning could potentially be a use case.) */
  hideLabel?: boolean;
  /** The select size should reflect the size of its content. */
  size?: 'small' | 'medium' | 'large' | 'full' | 'auto';
  disabled?: boolean;
  required?: boolean;
}

/** 
 * Combobox
 * 
 * https://www.w3.org/TR/wai-aria-practices/examples/combobox/combobox-select-only.html for accessibility implementation.
 * */
export default function Combobox({ 
    children,
    size = 'medium', 
    label, 
    error, 
    message, 
    hideLabel = false,
    disabled = false,
  }: Props) {

  const errorID = 'errorText';
  /* Add a space before the added class rather than inside the className attr on the tag. Looks cleaner. */
  let errorClass = error ? ' has-error' : '';
  let disabledClass = disabled ? ' is-disabled' : '';
  let hiddenClass = hideLabel ? ' aj-hidden' : '';

  const [menuActive, setMenuActive] = useState(false)

  const handleOpen = () => {
    if (menuActive) {
      setMenuActive(false)
    } else {
      setMenuActive(true)
    }
  }

  useEffect(() => {
    const closeMenu = () => {
      if (!menuActive) return
      setMenuActive(false)
    }

    window.addEventListener("click", closeMenu)

    return () => window.removeEventListener("click", closeMenu)
  }, [menuActive])

  return (
    <div className={`aj-dropdown is-${size}${errorClass}${disabledClass}`}>
      <label className={`aj-label${hiddenClass}`} id="comboLabel">
        {label}
        {message ?
          <p className="aj-label--message">{message}</p>
          : null
        }
      </label>
      <div className="aj-combobox">
        <div 
          className="aj-combobox__input is-searchable"
          aria-owns="listboxID"
          aria-expanded={menuActive}
          aria-haspopup="listbox"
          id="comboID"
          role="combobox"
          onClick={handleOpen}
        >
          <input
            type="text"
            aria-autocomplete='both'
            aria-controls='listboxID'
            aria-labelledby='comboLabel'
            aria-describedby={error ? errorID : ''}
            id="comboInputID"
          />
        </div>
        <ul
          className="aj-combobox__menu"
          role="listbox"
          id="lisboxID"
          aria-labelledby="comboLabel"
        >
          <li
            className="aj-combobox__option is-focused"
            role="option"
            id="op1"
          >
            Option
          </li>
          <li
            className="aj-combobox__option"
            role="option"
            id="op2"
          >
            {children}
          </li>
          <li
            className="aj-combobox__option"
            role="option"
            id="op3"
          >
            {children}
          </li>
        </ul>
      </div>
      {error ?
        <p id={errorID} className="aj-label--error">{error}</p>
        : null
      }
    </div>
  )
}
