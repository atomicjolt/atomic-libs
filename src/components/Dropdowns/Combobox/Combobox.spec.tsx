import React from "react";
import { render } from "@testing-library/react";
import Combobox from ".";

describe("matches snapshots", () => {
  const shared = {
    label: "Text Label",
    message: "message",
    error: "some error",
    value: "content",
    onChange: (value: any) => {},
    options: ["Option 1", "Option 2", "Option 3"],
  };

  it("matches default variant", () => {
    const result = render(<Combobox {...shared} />);
    expect(result.asFragment()).toMatchSnapshot();
  });

  it("matches floating variant", () => {
    const result = render(<Combobox variant="floating" {...shared} />);
    expect(result.asFragment()).toMatchSnapshot();
  });
});
