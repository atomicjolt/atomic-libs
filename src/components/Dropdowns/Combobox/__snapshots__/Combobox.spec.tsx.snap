// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="aje-dropdown is-medium has-error"
  >
    <label
      class="aje-label"
      for="combo-label-12345678900000000000"
    >
      Text Label
      <p
        class="aje-label--message"
      >
        message
      </p>
    </label>
    <div
      class="aje-combobox"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="combo-listbox-12345678900000000000"
        class="aje-combobox__input is-searchable"
        id="combo-combo-12345678900000000000"
        role="combobox"
      >
        <input
          aria-autocomplete="both"
          aria-controls="combo-listbox-12345678900000000000"
          aria-describedby="combo-error-12345678900000000000"
          aria-labelledby="combo-label-12345678900000000000"
          id="combo-input-12345678900000000000"
          type="text"
          value="content"
        />
      </div>
      <ul
        aria-labelledby="combo-label-12345678900000000000"
        class="aje-combobox__menu"
        id="combo-listbox-12345678900000000000"
        role="listbox"
      />
    </div>
    <p
      class="aje-label--error"
      id="combo-error-12345678900000000000"
    >
      some error
    </p>
  </div>
</DocumentFragment>
`;
